File: Example-ObjC/Embedded/CardfieldVC.m
//
//  ViewController.m
//  Embedded
//
//  Created by Sedlak, Stefan on 7/21/16.
//  Copyright © 2016 Wirecard Technologies GmbH. All rights reserved.
//

#import "CardfieldVC.h"

#import <libextobjc/EXTScope.h>
#import <WDeComCard/WDeComCard.h>
#import <WDeComCardScannerGallery/WDECCardFieldScannerGallery.h>

#define NSStringize_helper(x) #x
#define NSStringize(x) @NSStringize_helper(x)

#define AMOUNT [NSDecimalNumber decimalNumberWithMantissa:199 exponent:-2 isNegative:NO]

@interface CardfieldVC () <WDECCardFieldDelegate>

@property (weak, nonatomic) IBOutlet WDECCardFieldScannerGallery *cardField;
@property (weak, nonatomic) IBOutlet UILabel *cardFieldStateLabel;
@property (weak, nonatomic) IBOutlet UIButton *payBtn;

@end

@implementation CardfieldVC

- (void)viewDidLoad {
    [super viewDidLoad];
    [self.cardField becomeFirstResponder];
    self.cardField.manualCardBrandSelectionRequired = [[WDECCardLayout appearance] manualCardBrandSelectionRequired];
    [[WDECCardFieldScannerGallery appearance] setScanImage: [UIImage imageNamed:@"scan_bf2"]];
    [[WDECCardFieldScannerGallery appearance] setScanImageTintColor:[UIColor blueColor]];
    [[WDECCardFieldScannerGallery appearance] setEnableScanImageButton:YES];

    [[WDECCardFieldScannerGallery appearance] setScanToolbarButtonTitle:@"Scan Card"];
    [[WDECCardFieldScannerGallery appearance] setEnableScanToolbarButton:YES];
    
    //Local customization
    //self.cardField.scanImageTintColor = [UIColor blueColor];
    //self.cardField.scanImage = [UIImage imageNamed:@"scan_bf2"];
    //self.cardField.enableScanImageButton = YES;

    //self.cardField.scanToolbarButtonTitle = @"Scan Card";
    //self.cardField.enableScanToolbarButton = YES;
}

- (IBAction)onClear:(UIButton *)sender {
    [self.cardField clear];
}

- (IBAction)onPay:(UIButton *)sender {
    // if you create payment object just before calling makePayment you are sure that timestamp is correct
    WDECCardPayment *payment = [self createCardPayment];
    self.cardField.cardPayment = payment; // append card data
    payment = self.cardField.cardPayment;
    
    if (!self.client) {
        return;
    }
    sender.enabled = NO;
    @weakify(self, sender);
    [self.client makePayment:payment withCompletion:^(WDECPaymentResponse * _Nullable response, NSError * _Nullable error) {
        @strongify(self, sender);
        sender.enabled = YES;
        
        NSString *alertTitle = error ? @"Error" : @"Success";
        NSString *alertMessage = error ? error.localizedDescription : @"Item(s) has been purchased.";
        UIAlertController *ac = [UIAlertController alertControllerWithTitle:alertTitle
                                                                    message:alertMessage
                                                             preferredStyle:UIAlertControllerStyleAlert];
        
        [ac addAction:[UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:nil]];
        [self presentViewController:ac animated:YES completion:nil];
        
        // each request shall have unique ID, once it is used create new one
        self.cardField.cardPayment = [self createCardPayment];
    }];
}
#pragma mark - WDECCardFieldDelegate

- (void)cardField:(nonnull WDECCardField *)cardField didChangeState:(WDECCardFieldState)state {
    self.payBtn.enabled = cardField.valid;
    
#define SWITCH_CASE(value) case value: { text = NSStringize(value); break; }
    NSString *text = nil;
    switch (state) {
            SWITCH_CASE(WDECCardFieldStateCardInitial)
            SWITCH_CASE(WDECCardFieldStateJailbrokenDevice)
            SWITCH_CASE(WDECCardFieldStateCardValid)
            SWITCH_CASE(WDECCardFieldStateCardUnsupported)
            SWITCH_CASE(WDECCardFieldStateNumberEditting)
            SWITCH_CASE(WDECCardFieldStateNumberIncomplete)
            SWITCH_CASE(WDECCardFieldStateNumberInvalid)
            SWITCH_CASE(WDECCardFieldStateNumberValid)
            SWITCH_CASE(WDECCardFieldStateMonthEditting)
            SWITCH_CASE(WDECCardFieldStateYearEditting)
            SWITCH_CASE(WDECCardFieldStateExpirationDateIncomplete)
            SWITCH_CASE(WDECCardFieldStateExpirationDateInvalid)
            SWITCH_CASE(WDECCardFieldStateExpirationDateValid)
            SWITCH_CASE(WDECCardFieldStateSecurityCodeEditting)
            SWITCH_CASE(WDECCardFieldStateSecurityCodeIncomplete)
            SWITCH_CASE(WDECCardFieldStateSecurityCodeInvalid)
            SWITCH_CASE(WDECCardFieldStateSecurityCodeValid)
    }
#undef SWITCH_CASE
    
    self.cardFieldStateLabel.text = text;
}

- (WDECCardPayment *)createCardPayment {
    WDECCardPayment *payment = [WDECCardPayment new];
    payment.amount = AMOUNT;
    payment.currency = @"USD";
    payment.transactionType = WDECTransactionTypePurchase;
    static NSString *const WD_MERCHANT_ACCOUNT_ID = @"33f6d473-3036-4ca5-acb5-8c64dac862d1";
    static NSString *const WD_MERCHANT_SECRET_KEY = @"9e0130f6-2e1e-4185-b0d5-dc69079c75cc";
    [self merchant:WD_MERCHANT_ACCOUNT_ID signPayment:payment byMerchantSecretKey:WD_MERCHANT_SECRET_KEY];
    payment.accountHolder = [self accountHolder];
    
    return payment;
}

- (WDECCustomerData *)accountHolder {
    WDECCustomerData *accountHolder = nil;
    accountHolder = [WDECCustomerData new];
    accountHolder.lastName = @"Doe";
    
    return accountHolder;
}

@end

File: Example-Swift/Embedded/CardfieldVC.swift
//
//  ViewController.swift
//  Embedded
//
//  Created by Vrana, Jozef on 3/6/17.
//  Copyright © 2017 Vrana, Jozef. All rights reserved.
//

import WDeComCard
import WDeComCardScannerGallery

let AMOUNT = NSDecimalNumber.init(mantissa: 199, exponent: -2, isNegative: false)

class CardfieldVC: PaymemtVC, UIActionSheetDelegate, WDECCardFieldDelegate {
    
    @IBOutlet weak var cardField: WDECCardFieldScannerGallery!
    @IBOutlet weak var cardFieldStateLabel: UILabel!
    @IBOutlet weak var payBtn: UIButton!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.cardField.becomeFirstResponder()
        let cardLayout = WDECCardLayout.appearance()
        self.cardField.isManualCardBrandSelectionRequired = cardLayout.manualCardBrandSelectionRequired
        self.cardField.scanImage = UIImage(named: "scan_cf2")
        self.cardField.scanImageTintColor = UIColor.blue
        self.cardField.enableScanImageButton = true

        self.cardField.scanToolbarButtonTitle = "Scan Card"
        self.cardField.enableScanToolbarButton = true

        // Appearance posibility
        //WDECCardFieldScannerGallery.appearance().scanImage = UIImage(named: "scan_cf2")
        //WDECCardFieldScannerGallery.appearance().scanImageTintColor = UIColor.blue
        //WDECCardFieldScannerGallery.appearance().enableScanImageButton = true

        //WDECCardFieldScannerGallery.appearance().scanToolbarButtonTitle = "Scan Card"
        //WDECCardFieldScannerGallery.appearance().enableScanToolbarButton = true
    }
    
    @IBAction func onClearAction(_ sender: UIButton) {
        self.cardField.clear()
    }
    
    @IBAction func onPayAction(_ sender: UIButton) {
        // if you create payment object just before calling makePayment you are sure that timestamp is correct
        var payment = self.createCardPayment()
        self.cardField.cardPayment = payment // append card data
        
        payment = self.cardField.cardPayment
        self.client?.make(payment, withCompletion:{(response: WDECPaymentResponse?,error: Error?) in
            let alertTitle = error != nil ? "Error" : "Success"
            let alertMessage = error != nil ? error!.localizedDescription : "Item(s) has been purchased."
            let ac = UIAlertController(title: alertTitle, message: alertMessage, preferredStyle: .actionSheet)
            
            ac.addAction(UIAlertAction(title:"OK", style:.default, handler:nil));
            self.present(ac, animated:true, completion:nil)
            
            // each request shall have unique ID, once it is used create new one
            self.cardField.cardPayment = self.createCardPayment()
        })
    }
    
    func cardField(_ cardField: WDECCardField, didChange state: WDECCardFieldState) {
        self.payBtn.isEnabled = cardField.isValid
        
        var text : String?
        
        switch state {
        case WDECCardFieldState.cardInitial: text = "card initial"
        case WDECCardFieldState.jailbrokenDevice: text = "jailbroken device"
        case WDECCardFieldState.cardValid: text = "card valid"
        case WDECCardFieldState.cardUnsupported: text = "card unsupported"
        case WDECCardFieldState.numberEditting: text = "number editting"
        case WDECCardFieldState.numberIncomplete: text = "number incomplete"
        case WDECCardFieldState.numberInvalid: text = "number invalid"
        case WDECCardFieldState.numberValid: text = "number valid"
        case WDECCardFieldState.monthEditting: text = "month editting"
        case WDECCardFieldState.yearEditting: text = "year editting"
        case WDECCardFieldState.expirationDateIncomplete: text = "expiration date incomplete"
        case WDECCardFieldState.expirationDateInvalid: text = "expiration date invalid"
        case WDECCardFieldState.expirationDateValid: text = "expiration date valid"
        case WDECCardFieldState.securityCodeEditting: text = "security code editting"
        case WDECCardFieldState.securityCodeIncomplete: text = "security code incomplete"
        case WDECCardFieldState.securityCodeInvalid: text = "security code invalid"
        case WDECCardFieldState.securityCodeValid: text = "security code valid"
        }
        
        self.cardFieldStateLabel.text = text
    }
    
    func createCardPayment() -> WDECCardPayment {
        let payment = WDECCardPayment()
        payment.amount = AMOUNT
        payment.currency = "USD"
        payment.transactionType = .purchase
        let WD_MERCHANT_ACCOUNT_ID = "33f6d473-3036-4ca5-acb5-8c64dac862d1"
        let WD_MERCHANT_SECRET_KEY = "9e0130f6-2e1e-4185-b0d5-dc69079c75cc"
        self.merchantSignedPaymentByMerchantSecretKey(merchantAccountID: WD_MERCHANT_ACCOUNT_ID, payment: payment, merchantSecretKey: WD_MERCHANT_SECRET_KEY)
        let accountHolder = WDECCustomerData()
        accountHolder.firstName = "John"
        accountHolder.lastName = "Doe"
        payment.accountHolder = accountHolder
        
        return payment
    }
}

